packages/protocol/tipmanager/tipmanager.go

```diff
@@ -216,19 +180,165 @@ func (t *TipManager) selectTips(count int) (parents models.BlockIDs) {
 }
 
 // AllTips returns a list of all tips that are stored in the TipManger.
-func (t *TipManager) AllTips() []*scheduler.Block {
-	return t.tips.Keys()
+func (t *TipManager) AllTips() (allTips []*scheduler.Block) {
+	allTips = make([]*scheduler.Block, 0, t.tips.Size())
+	t.tips.ForEach(func(_ models.BlockID, value *scheduler.Block) bool {
+		allTips = append(allTips, value)
+		return true
+	})
+
+	return
 }
 
 // TipCount the amount of tips.
 func (t *TipManager) TipCount() int {
 	return t.tips.Size()
 }
 
+// FutureTipCount returns the amount of future tips per epoch.
+func (t *TipManager) FutureTipCount() (futureTipsPerEpoch map[epoch.Index]int) {
+	futureTipsPerEpoch = make(map[epoch.Index]int)
+	t.futureTips.ForEach(func(index epoch.Index, commitmentStorage *memstorage.Storage[commitment.ID, *memstorage.Storage[models.BlockID, *scheduler.Block]]) {
+		commitmentStorage.ForEach(func(cm commitment.ID, tipStorage *memstorage.Storage[models.BlockID, *scheduler.Block]) bool {
+			futureTipsPerEpoch[index] += tipStorage.Size()
+			return true
+		})
+	})
+
+	return
+}
+
+// PromoteFutureTips promotes to the main tippool all future tips that belong to the given commitment.
+func (t *TipManager) PromoteFutureTips(cm *commitment.Commitment) {
+	t.evictionMutex.RLock()
+	defer func() {
+		t.evictionMutex.RUnlock()
+		t.Evict(cm.Index())
+	}()
+
+	if futureEpochTips := t.futureTips.Get(cm.Index()); futureEpochTips != nil {
+		if tipsForCommitment, exists := futureEpochTips.Get(cm.ID()); exists {
+			tipsToPromote := make(map[models.BlockID]*scheduler.Block)
+			tipsForCommitment.ForEach(func(blockID models.BlockID, tip *scheduler.Block) bool {
+				tipsToPromote[blockID] = tip
+				return true
+			})
```

Comment:
```suggestion
			tipsForCommitment.ForEach(func(blockID models.BlockID, tip *scheduler.Block) bool {
				tipsToPromote[blockID] = tip
				
				for _, parentID := range tip.Parents() {
					if _, exists := tipsToPromote[parentID]; exists {
						// Do not add this tips parent because this is not a tip anymore
						tipsToNotPromote.Add(parentID)
					}
				}
				// Remove all tips referenced by all future tips valid for this commitment
				t.RemoveStrongParents(tip.ModelsBlock)
				
				return true
			})
			      
```

Reply:
This you cannot really do, because you are filling `tipsToPromote` while you are checking membership in it. So the end result will depend in what order the outer ForEach will iterate! Hence why we need to fill `tipsToPromote` entirely first.